pipeline/artifactory/artifactory.go

```diff
@@ -0,0 +1,429 @@
+				"mode":     v,
+			}).Error(err.Error())
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// runPipeForModeArchive uploads all ctx.Artifacts to instance
+func runPipeForModeArchive(ctx *context.Context, instance config.Artifactory) error {
+	sem := make(chan bool, ctx.Parallelism)
+	var g errgroup.Group
+
+	// Get all artifacts and upload them
+	for _, artifact := range ctx.Artifacts {
+		sem <- true
+		artifact := artifact
+		g.Go(func() error {
+			defer func() {
+				<-sem
+			}()
+
+			return uploadArchive(ctx, instance, artifact)
+		})
+	}
+
+	return g.Wait()
+}
+
+// uploadArchive will upload artifact in mode archive
+func uploadArchive(ctx *context.Context, instance config.Artifactory, artifact string) error {
+	var path = filepath.Join(ctx.Config.Dist, artifact)
+	return uploadAssetAndLog(ctx, instance, path, nil)
+}
+
+// uploadBinary will upload the current build and the current target in mode binary
+func uploadBinary(ctx *context.Context, instance config.Artifactory, build config.Build, target buildtarget.Target) error {
+	binary, err := getBinaryForUploadPerBuild(ctx, target)
+	if err != nil {
+		return err
+	}
+
+	return uploadAssetAndLog(ctx, instance, binary.Path, &target)
+}
+
+// uploadAssetAndLog uploads file to target and logs all actions
+func uploadAssetAndLog(ctx *context.Context, instance config.Artifactory, path string, target *buildtarget.Target) error {
+	secret := os.Getenv(fmt.Sprintf("ARTIFACTORY_%s_SECRET", strings.ToUpper(instance.Name)))
+
+	// Generate the target url
+	targetURL, err := resolveTargetTemplate(ctx, instance, target)
+	if err != nil {
+		msg := "artifactory: error while building the target url"
+		log.WithField("instance", instance.Name).WithError(err).Error(msg)
+		return errors.Wrap(err, msg)
+	}
+
+	// Handle the artifact
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close() // nolint: errcheck
+	_, name := filepath.Split(path)
+
+	// The target url needs to contain the artifact name
+	if !strings.HasSuffix(targetURL, "/") {
+		targetURL += "/"
+	}
+	targetURL += name
+
+	artifact, _, err := uploadAssetToArtifactory(ctx, targetURL, instance.Username, secret, file)
+	if err != nil {
+		msg := "artifactory: upload failed"
+		log.WithError(err).WithFields(log.Fields{
+			"instance": instance.Name,
+			"username": instance.Username,
+		}).Error(msg)
+		return errors.Wrap(err, msg)
+	}
+
+	log.WithFields(log.Fields{
+		"instance": instance.Name,
+		"mode":     instance.Mode,
+		"uri":      artifact.DownloadURI,
+	}).Info("uploaded successful")
+
+	return nil
+}
+
+// runPipeForModeBinary uploads all configured builds to instance
+func runPipeForModeBinary(ctx *context.Context, instance config.Artifactory, build config.Build) error {
+	sem := make(chan bool, ctx.Parallelism)
+	var g errgroup.Group
+
+	// Lets generate the build matrix, because we want
+	// to publish every target to Artifactory
+	for _, target := range buildtarget.All(build) {
```

Comment:
I think you don't need to do that... you can iterate over the `ctx.Binaries` directly :)

Reply:
I had a quick check and `ctx.Binaries` don't offer what the functionality represents.

`ctx.Binaries` looks like
```
map[string]map[string][]context.Binary{
    "darwinamd64":map[string][]context.Binary{
        "goreleaser_Darwin_x86_64":[]context.Binary{
            context.Binary{Name:"goreleaser", Path:"dist/goreleaser_Darwin_x86_64/goreleaser"}
        }
    }, 
    "linuxamd64":map[string][]context.Binary{
        "goreleaser_Linux_x86_64":[]context.Binary{
            context.Binary{Name:"goreleaser", Path:"dist/goreleaser_Linux_x86_64/goreleaser"}
        }
    },
...
```

and `buildtarget.All(build)` looks like
```
[]buildtarget.Target{
    buildtarget.Target{OS:"linux", Arch:"386", Arm:""}, 
    buildtarget.Target{OS:"linux", Arch:"amd64", Arm:""}, 
...
```

The reason for the `buildtarget.All(build)` usage is to be able to access the variables `OS`, `Arch` and `Arm` to replace them in the `target` template.
Those variables are not available in `ctx.Binaries`.

Do you share this? Or do i miss something?

Two options:
1. We keep the solution as is
2. We extend `ctx.Binaries` to extend the information

I prefer solution 1. for now (to get this functionality in and keep the PR "reviewable") and create a ticket for 2. to extend this feature and refactor this part here.
What do you think? I can create the tickets if you want.

Reply:
Ohh, true, I agree. 

This binaries and artifacts things in the context package are pretty messed and I'll rewrite them eventually (#455)

let's leave it as-is for now :)